src/write/encoder_string_writer.rs

```diff
@@ -0,0 +1,173 @@
+use crate::Config;
+use std::io;
+use std::io::Write;
+use super::encoder::EncoderWriter;
+
+/// A `Write` implementation that base64-encodes data using the provided config and accumulates the
+/// resulting base64 in memory, which is then exposed as a String via `into_inner()`.
+///
+/// # Examples
+///
+/// Buffer base64 in a new String:
+///
+/// ```
+/// use std::io::Write;
+///
+/// let mut enc = base64::write::EncoderStringWriter::new(base64::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // get the resulting String
+/// let b64_string = enc.into_inner();
+///
+/// assert_eq!("YXNkZg==", &b64_string);
+/// ```
+///
+/// Or, append to an existing String:
+///
+/// ```
+/// use std::io::Write;
+///
+/// let mut buf = String::from("base64: ");
+///
+/// let mut enc = base64::write::EncoderStringWriter::from(&mut buf, base64::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // release the &mut reference on buf
+/// let _ = enc.into_inner();
+///
+/// assert_eq!("base64: YXNkZg==", &buf);
+/// ```
+///
+/// # Panics
+///
+/// Calling `write()` (or related methods) or `finish()` after `finish()` has completed without
+/// error is invalid and will panic.
+///
+/// # Performance
+///
+/// Because it has to validate that the base64 is UTF-8, it is about 80% as fast as writing plain
+/// bytes to a `io::Write`.
+pub struct EncoderStringWriter<S: StrWrite> {
+    encoder: EncoderWriter<Utf8SingleCodeUnitWriter<S>>,
+}
+
+impl<S: StrWrite> EncoderStringWriter<S> {
+    /// Create a EncoderStringWriter that will append to the provided `StrWrite`.
+    pub fn from(str_writer: S, config: Config) -> Self {
+        EncoderStringWriter { encoder: EncoderWriter::new(Utf8SingleCodeUnitWriter { str_writer }, config) }
+    }
+
+    /// Encode all remaining buffered data, including any trailing incomplete input triples and
+    /// associated padding.
+    ///
+    /// Once this succeeds, no further writes or calls to this method are allowed.
+    ///
+    /// Returns the base64-encoded form of the accumulated written data.
+    pub fn into_inner(mut self) -> S {
+        self.encoder.finish()
+            .expect("Writing to a Vec<u8> should never fail")
+            .str_writer
+    }
+}
+
+impl EncoderStringWriter<String> {
+    /// Create a EncoderStringWriter that will encode into a new String with the provided config.
+    pub fn new(config: Config) -> Self {
+        EncoderStringWriter::from(String::new(), config)
+    }
+}
+
+impl <S: StrWrite> Write for EncoderStringWriter<S> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.encoder.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.encoder.flush()
+    }
+}
+
+/// An abstraction around infallible writes of `str`s.
+///
+/// Typically, this will just be String.
+pub trait StrWrite {
```

Comment:
better names?

Reply:
It reminds me of https://doc.rust-lang.org/std/iter/trait.Extend.html. I don't think you can use `Extend` directly though because I don't think it is implemented for `&mut String`.

You also might be able to use [`AsMut<String>`](https://doc.rust-lang.org/std/convert/trait.AsMut.html)? I think that would let you use `String` or `&String`?

Reply:
> You also might be able to use [`AsMut<String>`](https://doc.rust-lang.org/std/convert/trait.AsMut.html)? I think that would let you use `String` or `&String`?

Do you mean as a way to avoid implementing `StrWrite` for `&mut StrWrite`? I think I want that no matter what though so that it's not _only_ available for `String` and `&mut String`.

Reply:
No, I think it may avoid the need for the trait overall. Because you can use AsMut to convert `&mut String` and `&mut mut String` to `&mut String` which will just allow you to call `String.push_str` directly.

Reply:
Hmm, I'm having trouble getting the this to work.

Reply:
Ah, I had the wrong trait. You can use `BorrowMut`.

https://rust.godbolt.org/z/bPd9z9

Reply:
Good to know. I think I'll keep the StrWrite though so it's a step closer to being no_std friendly -- currently all that's needed to finish that effort would be to introduce an abstraction other than `Write` for `EncoderWriter`, AFAICT.